h2o-genmodel/src/main/java/hex/genmodel/easy/RowToRawDataConverter.java

```diff
@@ -66,110 +51,67 @@ public RowToRawDataConverter(GenModel m,
       }
-      else {
-        // Column has categorical value.
-        Object o = data.get(dataColumnName);
-        double value;
-        if (o instanceof String) {
-          String levelName = (String) o;
-          HashMap<String, Integer> columnDomainMap = _domainMap.get(index);
-          Integer levelIndex = columnDomainMap.get(levelName);
-          if (levelIndex == null) {
-            levelIndex = columnDomainMap.get(dataColumnName + "." + levelName);
-          }
-          if (levelIndex == null) {
-            if (_convertUnknownCategoricalLevelsToNa) {
-              value = Double.NaN;
-              _errorConsumer.unseenCategorical(dataColumnName, o, "Previously unseen categorical level detected, marking as NaN.");
-            } else {
-              _errorConsumer.dataTransformError(dataColumnName, o, "Unknown categorical level detected.");
-              throw new PredictUnknownCategoricalLevelException("Unknown categorical level (" + dataColumnName + "," + levelName + ")", dataColumnName, levelName);
-            }
-          }
-          else {
-            value = levelIndex;
+      if (Double.isNaN(value)) {
+        // If this point is reached, the original value remains NaN.
+        _errorConsumer.dataTransformError(columnName, o, "Given non-categorical value is unparseable, treating as NaN.");
+      }
+      rawData[targetIndex] = value;
+    } else {
+      // Column has categorical value.
+      if (o instanceof String) {
+        String levelName = (String) o;
+        if (! catEncoder.encodeCatValue(levelName, rawData)) {
+          if (_convertUnknownCategoricalLevelsToNa) {
+            catEncoder.encodeNA(rawData);
+            _errorConsumer.unseenCategorical(columnName, o, "Previously unseen categorical level detected, marking as NaN.");
+          } else {
+            _errorConsumer.dataTransformError(columnName, o, "Unknown categorical level detected.");
+            throw new PredictUnknownCategoricalLevelException("Unknown categorical level (" + columnName + "," + levelName + ")", columnName, levelName);
           }
-        } else if (o instanceof Double && Double.isNaN((double)o)) {
-          _errorConsumer.dataTransformError(dataColumnName, o, "Missing factor value detected, setting to NaN");
-          value = (double)o; //Missing factor is the only Double value allowed
-        } else {
-          _errorConsumer.dataTransformError(dataColumnName, o, "Unknown categorical variable type.");
-          throw new PredictUnknownTypeException(
-                  "Unexpected object type " + o.getClass().getName() + " for categorical column " + dataColumnName);
         }
-        rawData[index] = value;
+      } else if (o instanceof Double && Double.isNaN((double) o)) {
+        _errorConsumer.dataTransformError(columnName, o, "Missing factor value detected, setting to NaN");
+        catEncoder.encodeNA(rawData); // Missing factor is the only Double value allowed
```

Comment:
Javadoc should reflect the values allowed as well.

Reply:
The comment refers to the fact that at this point we know `o` is NaN, it says users cannot pass any other value into the RowData structure. However, this is not related to the `encodeNA` function - this function can do whatever and encode the value whichever way it wants (by default it will write NaN :))